### ISSUE TITLE
Add page categorization (or concept tagging) feature

### ISSUE THREAD

**Issue by virtuous-sloth** [gollum/gollum-lib#436]
This is a non-trivial feature request; implement MediaWiki's [Categories](https://www.mediawiki.org/wiki/Help:Categories) functionality which [TrueWiki](https://github.com/TrueBrain/TrueWiki) also implements. This would provide an alternative, more flexible and more dynamic indexing of pages in addition to using directory structures with breadcrumbs.

I assume YAML metadata could be used for the actual categorization of a page, with the actual categories (foo and bar here) being an isolated name space with as little restriction as possible (strings, basically) with no relation to any other naming, including directories. However, that names of the categories should be restricted to valid page names so that they themselves can be optionally instantiated with the custom category page content being rendered above the calculated index of categorized pages below. The custom content would also include the category page having the ability to have it's own categorization metadata in it's YAML, essentially allowing the creation of category hierarchies, i.e. sub-categorization.

The namespace isolation happens in MediaWiki using a namespace field in their database (normal pages having an implicit Page: namespace, uploaded files having a File: namespace, and categories having a Category: namespace.

TrueWiki, which also uses git as a backing store, solves this problem including the namespace in the directory path. The top two directory levels of the git repository are for language (e.g. /en/, /fr/) and namespace (/en/Page/, /fr/File/, etc.).

Gollum eschews the language and currently follows the default implicit page namespace (i.e. pages are any file with a markup path extension from the root on down) and maps the File: namespace to /uploads/. In this spirit, if the user marks a page's YAML with a category named foo then that category would have a red (not-yet-created) /categories/foo absolute path which would show all the pages in that category by default but allow the user to create a custom portion of the page which would be stored in /categories/foo.md, for example. This would include a mandatory hard-coded category called 'categories' which would render at the '/categories/categories' url path and allow for customization in a file at /categories/categories.md.

This is hard-coded in the sense that all category pages which render at /categories/somecategory path would implicitly have a metadata tag of "category: categories" regardless of whether there was a file /categories/somecategory.md with or without YAML frontmatter 'category: categories' or 'categories: [categories, foo, bar].

categories:
  - foo
  - bar
or
categories: [foo, bar]
or
category: foo
category: bar

Page rendering would include category membership much like breadcrumbs currently are, but at the bottom like with MediaWiki.

The advantage of categories over directory structure is that one can start page categorization arbitrarily by simply starting to assign arbitrary category names to pages along multiple categorization dimensions and find structure later; category pages themselves can be categorized, resulting in a hierarchical category structure, if desired, with subcategories. Also, pages can belong to multiple categories and therefore belong multiple category hierarchies (like dimensions in a categorization space), which is not possible with directory structure (which would still be available for one particular/different categorization).

Since this would provide an indexing function, the category index would have to generated on startup and stored in memory, which might cause performance problems, or persisted to a file and cached in memory, which adds complexity (where to store it, how to manage corruption and synchronization between the index file and the categorization source of truth, the YAML metadata).

**Comment by virtuous-sloth** [comment]
I think my above functional specification may not be clearest, so I assume improving it would be the first thing (I'd have) to do. It reads better in reference to the MediaWiki and TrueWiki specifications, so if we could avoid duplication and continue use them by reference, that would he helpful, I hope.

Although this feature request would be non-trivial, I do think it has the nice property of being fairly easy to isolate. Gollum already supports the YAML metadata with no restrictions, to my understanding. It already has the breadcrumbs as an example of boilerplate content rendered on every page.

**Comment by guillaume-d** [comment]
Hi!

What you initially said makes perfect sense to me and I am pretty much on the same page as you! :smile:

I implemented a _very early **prototype**_ a few months more or less along these lines, and finally took the time to minimally clean it up and publish it.


Some notes/caveats:

I use the term "tag" and not category as it seems to be more standard, see below.

This is and will stay Markdown-only unless someone else helps and maintains that part.

For tags I use YAML metadata.
Tag names must also be valid wiki page names (because each tag can also have a normal page), but are not otherwise limited.

There is no new concept of (tag) namespace, although the lists of tagged pages are stored for performance as `*..json` (sic!) files alongside the `*.md` files, but that's more to avoid file name collisions than for conceptual namespacing really.

The list of tagged pages is never displayed automatically anywhere _for now_, a [Gollum custom macro](https://github.com/gollum/gollum/wiki#custom-macros) must be invoked everywhere you want to have it displayed.
Accordingly there is no automatically generated tag pages, although I also agree it would be nice to have.

YAML metadata and thus tag names are still rendered as ugly and useless plain text (unchanged from Gollum's default), although that would be my next TODO.
Hence of course some server-side styling (or Javascript code?) to color not yet created tags is missing.
Fancy stuff like breadcrumbs and foldable subtag lists would be nice (given much more time).

Not thought about a `tags` tag yet, my implementation seems not to need it, but it might help as a marker to collect all _intended_ tags (those that are not spelling mistakes of real ones that is).

I decide to use `tags` in YAML and not `categories` as it already seems to have become quite standardized, see some research at https://gitlab.com/iguillaumed/playground/tagging/-/blob/main/tags.md
Is your 3rd YAML tag syntax proposal with repeated key name really valid YAML? That would be nice and my preferred syntax then!

As regards performance and other needed work, lots of things are still missing.
The most glaring omission is that tags deleted from pages are not deleted from the index pages yet.
And at least there should be a full reindexing command, if only as a 1st inefficient synchronization implementation.
Patches to the Gollum code itself may not be needed for what I want to achieve, also I still have to test how the current search implementation reacts to lots of YAML frontmatter...

So if you are interested, you can try and understand what I did there (it is not much code yet): https://gitlab.com/iguillaumed/playground/tagging-example/

**Comment by guillaume-d** [comment]
See also the earlier (2015) Gollum issue which among other things outlines a fully manual approach: https://github.com/gollum/gollum/issues/1058#issuecomment-132820707.
It might work for you if nothing else does, as it offers the most flexibility.

**Comment by guillaume-d** [comment]
> Hence of course some server-side styling (or Javascript code?) to color not yet created tags is missing.

Actually, rendering broken internal links in general (or "absent" pages as gollum-lib names it) is only supported when using the Gollum-specific double-bracketed link syntax (for example `[[NonExistingPage]]` ), see https://github.com/gollum/gollum-lib/issues/177#issuecomment-294149774! :-/
It seems to be implemented in https://github.com/gollum/gollum-lib/blob/master/lib/gollum-lib/filter/tags.rb.

As I only plan to support (standard) Markdown syntax, and as in my design tags also double as normal pages, only Javascript code seems possible for me. On the brighter side, it means that again no modification to Gollum will be needed, but just a [custom.js](https://github.com/gollum/gollum/wiki/Custom-Styling-and-Views#using-customjs-and-the---js-flag) instead.

**Comment by guillaume-d** [comment]
> YAML metadata and thus tag names are still rendered as ugly and useless plain text (unchanged from Gollum's default), although that would be my next TODO.

This is now implemented in the latest version of my PoC, see https://gitlab.com/iguillaumed/playground/tagging/-/blob/v0.0.3/CHANGELOG.md#003-2023-02-19: all present pages' tags are now hyperlinked to the associated tag pages.
This was done using [customized views](https://github.com/gollum/gollum/wiki/Custom-Styling-and-Views#customizing-views), modifying the small part of [`wiki_content.mustache`](https://github.com/gollum/gollum/blob/master/lib/gollum/templates/wiki_content.mustache) related to metadata.
The styling is very crude for now but could be made nicer using a [custom.css](https://github.com/gollum/gollum/wiki/Custom-Styling-and-Views#using-customcss-and-the---css-flag), some additional CSS classes may be required though.

**Comment by guillaume-d** [comment]
> [...] all present pages' tags are now hyperlinked to the associated tag pages. This was done using [customized views](https://github.com/gollum/gollum/wiki/Custom-Styling-and-Views#customizing-views), modifying the small part of [`wiki_content.mustache`](https://github.com/gollum/gollum/blob/master/lib/gollum/templates/wiki_content.mustache) related to metadata. [...]

I just implemented a more flexible (*) [Gollum custom macro](https://github.com/gollum/gollum/wiki#custom-macros) for that in addition to the above in the latest version of my PoC, see https://gitlab.com/iguillaumed/playground/tagging/-/blob/v0.0.4/CHANGELOG.md#004-2023-03-15.
The old mechanism still works for now but is now hidden behind Gollum's `display_metadata` flag in case someone still find it useful.

(*) Now the hyperlinked tag list can be positioned anywhere in the header and/or footer and/or sidebar and its display logic is unconstrained contrary to the powerless Mustache templates...

From there my next TODOs are probably: (re)indexing fixes/MVP-features, search tuning and cosmetic nice-to-have JS ergonomics (in that order).

**Comment by virtuous-sloth** [comment]
Thanks for all your work, Guillaume! I will hopefully try to test this out in the next couple weeks.

**Comment by guillaume-d** [comment]
> Fancy stuff like breadcrumbs and foldable subtag lists would be nice (given much more time).

For those interested I added minimal support for subtags (JS-less non-foldable tag tree) in the latest version of my PoC and fixed a few things, see https://gitlab.com/iguillaumed/playground/tagging/-/blob/v0.0.5/CHANGELOG.md#005-2023-04-10.

Please note another limitation I had not documented yet here or elsewhere: **_for the moment_ subdirectories are not supported** and do not work correctly: I want my users and myself to use tags for structuring.
I might (need to) fix that myself at some point but patches are welcome!
Also I am not sure I want tags to lie elsewhere than at the root anyway.

> Not thought about a `tags` tag yet, my implementation seems not to need it, but it might help as a marker to collect all _intended_ tags (those that are not spelling mistakes of real ones that is).

Doing the above I realized such a tag also allows to collect all hierarchies defined by a tag tree, although one could even define specialized tags so as to show only some of them. 

> From there my next TODOs are probably: (re)indexing fixes/MVP-features, search tuning and cosmetic nice-to-have JS ergonomics (in that order).

I now have subtags in a way which satisfies my current needs and I can live without other bells and whistles, so once (re)indexing and search is done as explained above I will probably declare victory :grinning: and move on to other things...
In the meantime feedback is welcome though.
